File: tests/layer2_filterbank/unit/test_utilities.py
"""
test_utilities.py
=================
Python port of:
    layer2_filterbank/unit/TestUtilities.m

Unit tests for format-conversion and length-inference utilities.

Covers: nonu2ucfmt, u2nonucfmt, filterbanklengthcoef, center_freqs.
"""

from __future__ import annotations

import pytest

import numpy as np

# ---------------------------------------------------------------------------
# filterbanklengthcoef
# ---------------------------------------------------------------------------

@pytest.mark.requires_impl
class TestFilterbanklengthcoefImpl:
    """TestUtilities: filterbanklengthcoef infers system length from coefficients."""

    def test_matches_filterbanklength(self, needs_impl):
        from cool_frames.filterbanks import filterbank, filterbanklengthcoef  # type: ignore
        from cool_frames.filters import (
            audfilters,  # type: ignore
            filterbanklength,  # type: ignore
        )
        Ls, fs = 1024, 8000
        g, a, fc, _, _info = audfilters(fs, Ls)
        L = filterbanklength(Ls, a)
        rng = np.random.default_rng(2)
        x = rng.standard_normal(Ls)
        c = filterbank(x, g, a)
        L_from_coef = filterbanklengthcoef(c, a)
        assert L_from_coef == L, \
            f"filterbanklengthcoef: got {L_from_coef}, expected {L}"

    def test_is_positive_integer(self, needs_impl):
        from cool_frames.filterbanks import filterbank, filterbanklengthcoef  # type: ignore
        from cool_frames.filters import audfilters  # type: ignore
        g, a, fc, _, _info = audfilters(8000, 1024)
        rng = np.random.default_rng(3)
        x = rng.standard_normal(1024)
        c = filterbank(x, g, a)
        L = filterbanklengthcoef(c, a)
        assert L > 0, "filterbanklengthcoef: output must be positive"
        assert int(L) == L, "filterbanklengthcoef: output must be an integer"

    def test_stereo_consistent(self, needs_impl):
        """Stereo input should give the same L as mono."""
        from cool_frames.filterbanks import filterbank, filterbanklengthcoef  # type: ignore
        from cool_frames.filters import audfilters  # type: ignore
        Ls = 1024
        g, a, fc, _, _info = audfilters(8000, Ls)
        rng = np.random.default_rng(4)
        c_stereo = filterbank(rng.standard_normal((Ls, 2)), g, a)
        c_mono   = filterbank(rng.standard_normal(Ls),     g, a)
        L_s = filterbanklengthcoef(c_stereo, a)
        L_m = filterbanklengthcoef(c_mono,   a)
        assert L_s == L_m, \
            f"filterbanklengthcoef: stereo gives L={L_s}, mono gives L={L_m}"


# ---------------------------------------------------------------------------
# center_freqs
# ---------------------------------------------------------------------------

@pytest.mark.requires_impl
class TestCentFreqsImpl:
    """TestUtilities: center_freqs returns normalised centre frequencies."""

    def test_length(self, needs_impl):
        from cool_frames.diagnostics import center_freqs  # type: ignore
        from cool_frames.filters import (
            audfilters,  # type: ignore
            filterbanklength,  # type: ignore
        )
        Ls = 1024
        g, a, fc, _, _info = audfilters(8000, Ls)
        L = filterbanklength(Ls, a)
        cfreq = center_freqs(g, L)
        assert len(cfreq) == len(g), \
            f"center_freqs: expected {len(g)} entries, got {len(cfreq)}"

    def test_normalised_range(self, needs_impl):
        """Normalised frequencies must lie in (-1, 1]."""
        from cool_frames.diagnostics import center_freqs  # type: ignore
        from cool_frames.filters import (
            audfilters,  # type: ignore
            filterbanklength,  # type: ignore
        )
        Ls = 1024
        g, a, fc, _, _info = audfilters(8000, Ls)
        L = filterbanklength(Ls, a)
        cfreq = np.asarray(center_freqs(g, L), dtype=float)
        assert np.all(np.abs(cfreq) <= 1.0 + 1e-12), \
            f"center_freqs: values out of (-1,1]: max|cf|={np.max(np.abs(cfreq)):.4f}"

    def test_monotone_interior(self, needs_impl):
        """Interior centre frequencies (excluding last boundary filter) are non-decreasing."""
        from cool_frames.diagnostics import center_freqs  # type: ignore
        from cool_frames.filters import (
            audfilters,  # type: ignore
            filterbanklength,  # type: ignore
        )
        Ls = 1024
        g, a, fc, _, _info = audfilters(8000, Ls)
        L = filterbanklength(Ls, a)
        cfreq = np.asarray(center_freqs(g, L), dtype=float)
        interior = cfreq[:-1]
        if len(interior) >= 2:
            assert np.all(np.diff(interior) > -1e-10), \
                "center_freqs: interior frequencies not non-decreasing"

    def test_consistent_with_audfilters_fc(self, needs_impl):
        """center_freqs (normalised) must correlate strongly with audfilters fc/Nyquist."""
        from cool_frames.diagnostics import center_freqs  # type: ignore
        from cool_frames.filters import (
            audfilters,  # type: ignore
            filterbanklength,  # type: ignore
        )
        Ls, fs = 1024, 8000
        g, a, fc, _, _info = audfilters(fs, Ls)
        L = filterbanklength(Ls, a)
        cfreq   = np.asarray(center_freqs(g, L), dtype=float)
        fc_norm = np.asarray(fc, dtype=float) / (fs / 2)
        M = len(g)
        interior = slice(1, M - 1)
        if M - 2 >= 2:
            corr_matrix = np.corrcoef(cfreq[interior], fc_norm[interior])
            assert corr_matrix[0, 1] > 0.99, \
                f"center_freqs: low correlation {corr_matrix[0,1]:.4f} with audfilters fc"


# ---------------------------------------------------------------------------
# pack_coefficients / unpack_coefficients (dense+mask batching)
# ---------------------------------------------------------------------------

class TestPackCoefficients:
    def test_roundtrip_nonuniform(self):
        import numpy as np
        from cool_frames.filterbanks import filterbank, pack_coefficients, unpack_coefficients
        from cool_frames.filters import audfilters
        g, a, fc, L, _info = audfilters(8000, 4000)
        c = filterbank(np.random.default_rng(0).standard_normal(L), g, a, L)
        dense, mask = pack_coefficients(c)
        assert dense.shape == (len(c), max(np.asarray(cm).size for cm in c))
        assert dense.shape == mask.shape
        c2 = unpack_coefficients(dense, mask)
        assert len(c2) == len(c)
        assert all(np.array_equal(np.asarray(c[m]).ravel(), c2[m]) for m in range(len(c)))

    def test_mask_counts_match_lengths(self):
        import numpy as np
        from cool_frames.filterbanks import pack_coefficients
        c = [np.arange(3.0), np.arange(5.0), np.arange(1.0)]
        dense, mask = pack_coefficients(c)
        assert dense.shape == (3, 5)
        assert list(mask.sum(axis=1)) == [3, 5, 1]

File: tests/layer2_filterbank/unit/TestUtilities.m
classdef TestUtilities < matlab.unittest.TestCase
%TESTUTILITIES  Unit tests for utility entry points.
%
%   Covers: nonu2ucfmt, u2nonucfmt, filterbanklengthcoef, cent_freqs.

    properties
        sig
        p
        g       % ERB filters
        a       % subsampling
        fc      % center frequencies (Hz)
        L       % system length
        M       % filter count
        % nonu2u conversion data (built once)
        gu      % uniform equivalent filters
        au      % uniform hop
        p_vec   % copies-per-channel vector
        c       % non-uniform coefficients from filterbank
        cu      % uniform coefficients from nonu2ucfmt
    end

    methods (TestClassSetup)
        function setupClass(tc)
            sharedDir = fullfile(fileparts(mfilename('fullpath')), '..', '..', 'shared');
            addpath(sharedDir);
            setup_filterbank_paths();
            [tc.sig, tc.p] = make_test_params();

            [tc.g, tc.a, tc.fc, tc.L] = audfilters(tc.p.fs, tc.p.Ls);
            tc.M = numel(tc.g);

            % Build uniform equivalent bank and coefficient format.
            [tc.gu, tc.au, tc.p_vec] = nonu2ufilterbank(tc.g, tc.a);
            tc.c  = filterbank(tc.sig.noise_mono, tc.g, tc.a);
            tc.cu = nonu2ucfmt(tc.c, tc.p_vec);
        end
    end

    % ── nonu2ucfmt ────────────────────────────────────────────────────────
    methods (Test)

        function testNonu2ucfmtOutputIsCell(tc)
            tc.verifyTrue(iscell(tc.cu), ...
                'nonu2ucfmt: output must be a cell array.');
        end

        function testNonu2ucfmtOutputLength(tc)
            tc.verifyEqual(numel(tc.cu), sum(tc.p_vec), ...
                'nonu2ucfmt: output must have sum(p) cells.');
        end

        function testNonu2ucfmtSubchannelColumns(tc)
            % Each uniform sub-channel must have 1 column (mono input).
            for k = 1 : numel(tc.cu)
                tc.verifyEqual(size(tc.cu{k}, 2), 1, ...
                    sprintf('nonu2ucfmt: cu{%d} must have 1 column for mono input.', k));
            end
        end

    end

    % ── u2nonucfmt ────────────────────────────────────────────────────────
    methods (Test)

        function testU2nonucfmtRoundtrip(tc)
            % nonu2ucfmt followed by u2nonucfmt must recover the original.
            c2 = u2nonucfmt(tc.cu, tc.p_vec);
            tc.verifyEqual(numel(c2), tc.M, ...
                'u2nonucfmt roundtrip: output must have M channels.');
            err = zeros(1, tc.M);
            for m = 1 : tc.M
                denom = norm(tc.c{m}, 'fro') + eps;
                err(m) = norm(c2{m} - tc.c{m}, 'fro') / denom;
            end
            tc.verifyLessThan(max(err), 1e-10, ...
                'u2nonucfmt(nonu2ucfmt(c, p), p) must recover c exactly.');
        end

        function testU2nonucfmtOutputIsCell(tc)
            c2 = u2nonucfmt(tc.cu, tc.p_vec);
            tc.verifyTrue(iscell(c2), ...
                'u2nonucfmt: output must be a cell array.');
        end

        function testU2nonucfmtOutputLength(tc)
            c2 = u2nonucfmt(tc.cu, tc.p_vec);
            tc.verifyEqual(numel(c2), tc.M, ...
                'u2nonucfmt: output must have M channels (= numel(p)).');
        end

    end

    % ── filterbanklengthcoef ──────────────────────────────────────────────
    methods (Test)

        function testFilterbanklengthcoefMatchesFilterbanklength(tc)
            % The length inferred from the coefficients must equal
            % the length reported by filterbanklength.
            L_from_coef = filterbanklengthcoef(tc.c, tc.a);
            tc.verifyEqual(L_from_coef, tc.L, ...
                'filterbanklengthcoef: must return the same L as filterbanklength.');
        end

        function testFilterbanklengthcoefIsPositiveInteger(tc)
            L_fc = filterbanklengthcoef(tc.c, tc.a);
            tc.verifyGreaterThan(L_fc, 0, ...
                'filterbanklengthcoef: output must be positive.');
            tc.verifyEqual(L_fc, round(L_fc), ...
                'filterbanklengthcoef: output must be an integer.');
        end

        function testFilterbanklengthcoefStereoConsistent(tc)
            % Stereo input should give the same length as mono.
            c_stereo = filterbank(tc.sig.noise_stereo, tc.g, tc.a);
            L_stereo = filterbanklengthcoef(c_stereo, tc.a);
            tc.verifyEqual(L_stereo, tc.L, ...
                'filterbanklengthcoef: stereo input must give the same L as mono.');
        end

    end

    % ── cent_freqs ────────────────────────────────────────────────────────
    methods (Test)

        function testCentFreqsLength(tc)
            cfreq = cent_freqs(tc.g, tc.L);
            tc.verifyEqual(numel(cfreq), tc.M, ...
                'cent_freqs: output length must equal number of filters.');
        end

        function testCentFreqsInNormalisedRange(tc)
            cfreq = cent_freqs(tc.g, tc.L);
            tc.verifyTrue(all(abs(cfreq) <= 1 + 1e-12), ...
                'cent_freqs: all values must lie in (-1, 1].');
        end

        function testCentFreqsMonotonic(tc)
            % For the ERB filter bank, center frequencies should be ordered.
            % The last (highpass) boundary filter is excluded: its circular centre
            % of gravity wraps past the Nyquist bin and maps to a negative value,
            % which is a known artefact of the periodic DFT, not an ordering bug.
            % (testCentFreqsMatchAudfiltersNormalized already verifies the interior
            % filters are consistent with audfilters fc, using the same exclusion.)
            cfreq = cent_freqs(tc.g, tc.L);
            cfreq_interior = cfreq(1 : tc.M - 1);
            if numel(cfreq_interior) >= 2
                tc.verifyTrue(all(diff(cfreq_interior) > -1e-10), ...
                    'cent_freqs: center frequencies must be non-decreasing (excluding last boundary filter).');
            end
        end

        function testCentFreqsMatchAudfiltersNormalized(tc)
            % Normalised cent_freqs should equal fc_Hz / (fs/2).
            cfreq      = cent_freqs(tc.g, tc.L);
            fc_norm    = tc.fc / (tc.p.fs / 2);
            % The two frequency vectors should be consistent (correlation > 0.99).
            % We compare the interior filters only, avoiding DC / Nyquist edge filters.
            interior   = 2 : tc.M - 1;
            if numel(interior) >= 2
                r = corrcoef(cfreq(interior), fc_norm(interior));
                tc.verifyGreaterThan(r(1,2), 0.99, ...
                    'cent_freqs: must be consistent with audfilters fc (correlation check).');
            end
        end

    end

end
